File: hello_world/android/src/main/kotlin/dev/buijs/klutter/cookbook/examples/hello_world/HelloWorldPlugin.kt
package dev.buijs.klutter.cookbook.examples.hello_world

import android.app.Activity
import android.content.Context
import dev.buijs.klutter.*
import dev.buijs.klutter.EventChannelFacade
import dev.buijs.klutter.MethodChannelFacade
import dev.buijs.klutter.registerEventSink
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import dev.buijs.klutter.cookbook.examples.hello_world.platform.*

private val methodChannelNames = setOf(
        "dev.buijs.klutter.cookbook.examples.hello_world/channel/hello_world",
)

private val eventChannelNames = emptySet<String>()


class HelloWorldPlugin: FlutterPlugin, MethodCallHandler, StreamHandler, ActivityAware {

    private val mainScope = CoroutineScope(Dispatchers.Main)
    private lateinit var activity: Activity
    private lateinit var mcFacade: MethodChannelFacade
    private lateinit var ecFacade: EventChannelFacade

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        this.mcFacade = MethodChannelFacade(this, binding.binaryMessenger, methodChannelNames)
        this.ecFacade = EventChannelFacade(this, binding.binaryMessenger, eventChannelNames)
    }        

    override fun onMethodCall(call: MethodCall, result: Result) {
        mainScope.launch {
            onEvent(
                event = call.method,
                data = call.arguments,
                result = result
            )
        }
    }


    override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink) {
        when (arguments) {
            else -> {
                eventSink.error("Unknown topic", "$arguments", null)
            }
        }
    }

    override fun onCancel(arguments: Any?) {
    }

    override fun onDetachedFromEngine(
        binding: FlutterPlugin.FlutterPluginBinding
    ) {
    }

    override fun onAttachedToActivity(
        binding: ActivityPluginBinding
    ) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(
        binding: ActivityPluginBinding
    ) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        // nothing
    }

    override fun onDetachedFromActivityForConfigChanges() {
        // nothing
    }

    suspend fun <T> onEvent(event: String, data: T?, result: Result) { 
           try {
               when(event) {
                "welcomeMessage" ->
                    result.success(HelloWorld().sayWelcomeUser(data as String?))
                "whereAmI" ->
                    result.success(HelloWorld().askWhereAmI())
                   else -> result.notImplemented()
               }
           } catch(e: Exception) {
               result.error("10101", e.message, e.stackTrace)
           }
       }
}

File: hello_world/ios/Classes/HelloWorldPlugin.swift
import Flutter
import UIKit
import Platform

public class HelloWorldPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    static let mcs: [String] = [
      "dev.buijs.klutter.cookbook.examples.hello_world/channel/hello_world", 
    ]

    static let ecs: [String] = [
    ]

    var methodChannels: Set<FlutterMethodChannel> = []
    var eventChannels: Set<FlutterEventChannel> = []


    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let instance = HelloWorldPlugin()
        
        for name in mcs {
           let channel = FlutterMethodChannel(name: name, binaryMessenger: messenger)
            instance.methodChannels.insert(channel)
            registrar.addMethodCallDelegate(instance, channel: channel)
        }
        
        for name in ecs {
            let channel = FlutterEventChannel(name: name, binaryMessenger: messenger)
            channel.setStreamHandler(instance)
        }
    }     
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let data = call.arguments
        switch call.method {
        case "whereAmI":
            self.whereAmI(data: data, result: result)
        case "welcomeMessage":
            self.welcomeMessage(data: data, result: result)
         default:
            result(FlutterMethodNotImplemented)
         }
     }

    func whereAmI(data: Any?, result: @escaping FlutterResult) {
       let value = HelloWorld().askWhereAmI()
        result(value)
    }
    
    func welcomeMessage(data: Any?, result: @escaping FlutterResult) {
        let dataOrNull = TypeHandlerKt.stringOrNull(data: data)
        if(dataOrNull == nil) {
           result(FlutterError(code: "ERROR_CODE",
                           message: "Unexpected data type: \(dataOrNull)",
                           details: nil))
        } else {
           result(HelloWorld().sayWelcomeUser(username: (dataOrNull as! String)))
         }
     }
    public func onListen(withArguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        let topic = withArguments ?? "none"
        switch "\(topic)" {          case "none":
           eventSink(FlutterError(code: "ERROR_CODE",
                                         message: "Topic not provided!",
                                         details: ""))
        default:
           eventSink(FlutterError(code: "ERROR_CODE",
                               message: "Unknown topic",
                               details: "\(topic)"))
        }
        return nil
     }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        for channel in eventChannels {
            channel.setStreamHandler(nil)
        }
        return nil
    }

}